### PR TITLE
[CLOUD-270] Sync private repositories through GitHub App

### PR DESCRIPTION
**Pull request by unknwon**
Due to the fact that the code host connection created by the GitHub App does not currently store the access token in the config itself (which will be addressed separately with more care in CLOUD-255), we need to exchange an installation access token for our GitHub App in order to compose the repository clone URLs for repositories that are synced through GitHub App, at the time of cloning and updating.

Notes:

1. I extracted the previously embedded function body of `GetRemoteURLFunc` to its own function (`getRemoteURLFunc`) in order to write tests for it.
2. I am aware of [similar boilerplates in other places already](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+:%3D+auth.NewOAuthBearerTokenWithGitHubApp%28&patternType=literal), but I choose to leave it for CLOUD-255 because I think it is currently unclear how to best utilize the code reusability within the scope of this PR.

## Test plan

1. Use our [GitHub App local testing guide](https://docs.google.com/document/d/1FYTIPLsT4CAT0jPHoCbTb7JcS8dySX4am_rmQZlVJVE/edit) to set up the organization code host connection, allow access to at least one public and one private repositories for the installation.
2. Pick at least one public and one private repositories to be synced into Sourcegraph.
3. All selected repositories should be available for search once the initial cloning is done.

### CODE REVIEW

File: cmd/gitserver/main.go [hunk 1]
```diff
@@ -284,6 +277,94 @@ func getDB() (*sql.DB, error) {
...
+				if err != nil {
+					return "", errors.Wrap(err, "parse installation ID")
+				}
+
+				ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+				defer cancel()
+
+				// TODO(cloud-saas): Cache the installation access token until it expires, see
+				// https://sourcegraph.atlassian.net/browse/CLOUD-255
+				token, err := client.CreateAppInstallationAccessToken(ctx, installationID)
```
Comment (ryanslade) [inline cmd/gitserver/main.go:346]: Not having caching in place makes me slightly nervous as I think try and get the remote URL quite often and each time will be making an external API call to GitHub

**Review by ryanslade — APPROVED** [review]

File: cmd/gitserver/main.go (hunk 1 above)
Comment (david-sandy) [reply cmd/gitserver/main.go:346]: Would you please include the @sourcegraph/security team when you design the caching of installation access tokens?  There are some security features we will need in place to ensure these tokens are safe.

Comment (unknwon) [reply cmd/gitserver/main.go:346]: > Would you please include the @sourcegraph/security team when you design the caching of installation access tokens? There are some security features we will need in place to ensure these tokens are safe.

@david-sandy sure! That said, the token will be stored within external service config which is already an encrypted DB column.

**Comment by unknwon** [comment]
Merging as-is, happy to address any post-merge comments!

**Pull request merged**